earth2grid/_regrid.py

```diff
@@ -48,22 +84,147 @@ def forward(self, x):
+        x_query = x_query.double()
+        y_query = y_query.double()
+
+        if torch.any(x_coords[1:] < x_coords[:-1]):
+            raise ValueError("x_coords must be in non-decreasing order.")
+
+        if torch.any(y_coords[1:] < y_coords[:-1]):
+            raise ValueError("y_coords must be in non-decreasing order.")
+
+        # Find indices for the closest lower and upper bounds in x and y directions
+        x_l_idx = torch.searchsorted(x_coords, x_query, right=True) - 1
+        x_u_idx = x_l_idx + 1
+        y_l_idx = torch.searchsorted(y_coords, y_query, right=True) - 1
+        y_u_idx = y_l_idx + 1
+
+        # fill in nan outside mask
+        def isin(x, a, b):
+            return (x <= b) & (x >= a)
+
+        mask = (
+            isin(x_l_idx, 0, x_coords.size(0) - 2)
+            & isin(x_u_idx, 1, x_coords.size(0) - 1)
+            & isin(y_l_idx, 0, y_coords.size(0) - 2)
+            & isin(y_u_idx, 1, y_coords.size(0) - 1)
+        )
+        x_u_idx = x_u_idx[mask]
+        x_l_idx = x_l_idx[mask]
+        y_u_idx = y_u_idx[mask]
+        y_l_idx = y_l_idx[mask]
+        x_query = x_query[mask]
+        y_query = y_query[mask]
+
+        # Compute weights
+        x_l_weight = (x_coords[x_u_idx] - x_query) / (x_coords[x_u_idx] - x_coords[x_l_idx])
+        x_u_weight = (x_query - x_coords[x_l_idx]) / (x_coords[x_u_idx] - x_coords[x_l_idx])
+        y_l_weight = (y_coords[y_u_idx] - y_query) / (y_coords[y_u_idx] - y_coords[y_l_idx])
+        y_u_weight = (y_query - y_coords[y_l_idx]) / (y_coords[y_u_idx] - y_coords[y_l_idx])
+        weights = torch.stack(
+            [x_l_weight * y_l_weight, x_u_weight * y_l_weight, x_l_weight * y_u_weight, x_u_weight * y_u_weight], dim=-1
+        )
+
+        stride = x_coords.size(-1)
+        index = torch.stack(
+            [
+                x_l_idx + stride * y_l_idx,
+                x_u_idx + stride * y_l_idx,
+                x_l_idx + stride * y_u_idx,
+                x_u_idx + stride * y_u_idx,
+            ],
+            dim=-1,
+        )
+        self.register_buffer("weights", weights)
+        self.register_buffer("mask", mask)
+        self.register_buffer("index", index)
+
+    def forward(self, z: torch.Tensor):
+        """
+        Interpolate the field
+
+        Args:
+            z: shape [Y, X]
+        """
+        *shape, y, x = z.shape
+        zrs = z.view(-1, y * x).T
+        # using embedding bag is 2x faster on cpu and 4x on gpu.
+        output = torch.nn.functional.embedding_bag(self.index, zrs, per_sample_weights=self.weights, mode='sum')
+        interpolated = torch.full(
+            [self.mask.numel(), zrs.shape[1]], fill_value=self.fill_value, dtype=z.dtype, device=z.device
+        )
+        interpolated.masked_scatter_(self.mask.unsqueeze(-1), output)
+        interpolated = interpolated.T.view(*shape, self.mask.numel())
+        return interpolated
+
+
+def S2NearestNeighborInterpolator(
```

Comment:
Name might be a bit misleading with the unification? Might be better to call it BarycentricInterpolator and can maybe have NearestNeighbor be a specialization of this with `k = 1` or just note this `k = 1` behavior in the docs?

Reply:
I realized it's not actually barycentric renaming to `KNNInterpolatorS2`.

Reply:
new name is actually `KNNS2Interpolator`